python/ray/experimental/gpu_object_manager/gpu_object_store.py

```diff
@@ -122,23 +152,85 @@ def add_gpu_object(
             gpu_object: A list of tensors representing the GPU object.
             is_primary: Whether the GPU object is the primary copy.
         """
-        if is_primary:
-            self.primary_gpu_object_ids.add(obj_id)
-        self.gpu_object_store[obj_id] = gpu_object
+        with self.object_present_cv:
+            if is_primary:
+                self.primary_gpu_object_ids.add(obj_id)
+            self.gpu_object_store[obj_id] = gpu_object
+            self.object_present_cv.notify_all()
 
     def is_primary_copy(self, obj_id: str) -> bool:
-        return obj_id in self.primary_gpu_object_ids
+        with self.lock:
+            return obj_id in self.primary_gpu_object_ids
+
+    def wait_and_get_object(
+        self, obj_id: str, timeout: Optional[float] = None
+    ) -> List["torch.Tensor"]:
+        """Atomically waits for the GPU object to be present in the GPU object
+        store, then gets it.  If the object is not present after the optional
+        timeout, raise a TimeoutError.
+
+        Args:
+            obj_id: The object ID to wait for.
+            timeout: The maximum time in seconds to wait for the object to be
+            present in the GPU object store. If not specified, wait
+            indefinitely.
 
-    def remove_gpu_object(self, obj_id: str):
+        Returns:
+            The tensors in the GPU object.
         """
-        Remove the GPU object from the GPU object store.
+        with self.lock:
+            self._wait_object(obj_id, timeout)
```

Comment:
`_wait_object` will return the object. We don't need to call `get_object`.

Reply:
Thanks, let me update _wait_object to not return the object. That way we always use get_object to read.